cli/src/claim_invitation.rs

```diff
@@ -0,0 +1,219 @@
+        DeviceClaimInProgress2Ctx, DeviceClaimInProgress3Ctx, DeviceClaimInitialCtx,
+        UserClaimFinalizeCtx, UserClaimInProgress1Ctx, UserClaimInProgress2Ctx,
+        UserClaimInProgress3Ctx, UserClaimInitialCtx, UserOrDeviceClaimInitialCtx,
+    },
+    BackendInvitationAddr, ClientConfig, DeviceAccessStrategy,
+};
+
+use crate::utils::*;
+
+#[derive(Args)]
+pub struct ClaimInvitation {
+    /// Parsec config directory
+    #[arg(short, long, default_value_os_t = get_default_config_dir())]
+    config_dir: PathBuf,
+    /// Device slughash
+    #[arg(short, long)]
+    device: Option<String>,
+    /// Server invitation address (e.g.: parsec://127.0.0.1:41905/Org?no_ssl=true&action=claim_user&token=4e45cc21e7604af196173ff6c9184a1f)
+    #[arg(short, long)]
+    addr: BackendInvitationAddr,
+}
+
+pub async fn claim_invitation(claim_invitation: ClaimInvitation) -> anyhow::Result<()> {
+    let ctx = step0(claim_invitation.addr).await?;
+
+    match ctx {
+        UserOrDeviceClaimInitialCtx::User(ctx) => {
+            let ctx = step1_user(ctx).await?;
+            let ctx = step2_user(ctx).await?;
+            let ctx = step3_user(ctx).await?;
+            let ctx = step4_user(ctx).await?;
+            save_user(ctx).await
+        }
+        UserOrDeviceClaimInitialCtx::Device(ctx) => {
+            let ctx = step1_device(ctx).await?;
+            let ctx = step2_device(ctx).await?;
+            let ctx = step3_device(ctx).await?;
+            let ctx = step4_device(ctx).await?;
+            save_device(ctx).await
+        }
+    }
+}
+
+/// Step 0: retrieve info
+async fn step0(addr: BackendInvitationAddr) -> anyhow::Result<UserOrDeviceClaimInitialCtx> {
+    let handle = start_spinner("Retrieving invitation info");
+
+    let ctx = claimer_retrieve_info(Arc::new(ClientConfig::default().into()), addr).await?;
+
+    handle.done();
+
+    Ok(ctx)
+}
+
+/// Step 1: wait peer
+async fn step1_user(ctx: UserClaimInitialCtx) -> anyhow::Result<UserClaimInProgress1Ctx> {
+    println!(
+        "Invitation greeter: {YELLOW}{}{RESET}",
+        ctx.greeter_human_handle()
+    );
+
+    let handle = start_spinner("Initializing connection with greeter for claiming user");
+
+    let ctx = ctx.do_wait_peer().await?;
+
+    handle.done();
+
+    Ok(ctx)
+}
+
+/// Step 1: wait peer
+async fn step1_device(ctx: DeviceClaimInitialCtx) -> anyhow::Result<DeviceClaimInProgress1Ctx> {
+    println!(
+        "Invitation greeter: {YELLOW}{}{RESET}",
+        ctx.greeter_human_handle()
+    );
+
+    let handle = start_spinner("Initializing connection with greeter for claiming device");
+
+    let ctx = ctx.do_wait_peer().await?;
+
+    handle.done();
+
+    Ok(ctx)
+}
+
+/// Step 2: signify trust
+async fn step2_user(ctx: UserClaimInProgress1Ctx) -> anyhow::Result<UserClaimInProgress2Ctx> {
+    let mut input = String::new();
+    let sas_codes = ctx.generate_greeter_sas_choices(3);
```

Comment:
On parsec-2.16 we display 4 code, any reason why we reduce it to 3 ?

Reply:
Really ?
I was in old 2.16 branch

Reply:
See: https://github.com/Scille/parsec-cloud/blob/releases/2.16/parsec/core/cli/invitation.py#L434

Reply:
We display 4 codes in the GUI ;)

https://github.com/Scille/parsec-cloud/blob/6a25d0e49c511de0c6aa8ea27a42e0039c1ea48f/libparsec/crates/types/src/invite.rs#L135-L153

From the implementation we display N+1 codes

Reply:
I know it's 4 in GUI, but in CLI it is 3 :)

Reply:
```
while sas_codes.len() < size
```
Yes :sweat_smile: it's 3